### PR TITLE
[Misc] Add fully interleaved support for multimodal 'string' content format

### LINKED ISSUE #12885: [Feature]: Fully interleaved support for multimodal prompts
### 🚀 The feature, motivation and pitch

Hello! Firstly, thank you for such a wonderful library.

I am currently developing a custom prompt auto-optimization tool that requires support for multimodal multi-turn conversations. As part of the optimization prompt for teacher model, there will be a section containing examples of student model request processing. The structure of these examples will look something like this:
```
## Example N ##
<message_ID>__<var_name> = <var_value>
...
<message_ID>__<var_name> = <var_value>

Predicted answer: <pred>
Ground truth answer: <gtruth>
Conclusion: <conclusion>
```

Here: 
* Each message can be of any role (system, user, assistant)
* Variables can represent either text or image data

From my understanding of the code [here](https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/chat_utils.py#L676) 
image tokens are currently appended to the beginning of the text prompt
```python
def _get_full_multimodal_text_prompt(placeholder_counts: Dict[str, int],
                                     text_prompt: str) -> str:
    """Combine multimodal prompts for a multimodal language model."""
    ...
    ...
    return "\n".join(missing_placeholders + [text_prompt])
```

This approach stacks all image tokens together at the start of the prompt, which could potentially impact the performance of the teacher model. I would like to know if it is possible to implement fully interleaved support for multimodal prompts?

I have implemented some custom modifications to the code, but I am uncertain whether these changes are compatible with all VLM architectures or if they might affect subsequent processing stages:

```python
IMAGE_PLACEHOLDER = "<##IMAGE##>"

..............

def _get_full_multimodal_text_prompt(placeholder_counts: Dict[str, int],
                                     texts: List[str],
                                     interleave: bool) -> str:
    """Combine multimodal prompts for a multimodal language model."""
    return (
        _full_multimodal_text_prompt_interleave(placeholder_counts, texts) if interleave else 
        _full_multimodal_text_prompt_simple(placeholder_counts, "\n".join(texts))
        )


def _full_multimodal_text_prompt_simple(placeholder_counts: Dict[str, int],
                                        text_prompt: str) -> str:
    # Look through the text prompt to check for missing placeholders
    missing_placeholders: List[str] = []
    for placeholder in placeholder_counts:

        # For any existing placeholder in the text prompt, we leave it as is
        placeholder_counts[placeholder] -= text_prompt.count(placeholder)

        if placeholder_counts[placeholder] < 0:
            raise ValueError(
                f"Found more '{placeholder}' placeholders in input prompt than "
                "actual multimodal data items.")

        missing_placeholders.extend([placeholder] *
                                    placeholder_counts[placeholder])

    # NOTE: For now we always add missing placeholders at the front of
    # the prompt. This may change to be customizable in the future.
    return "\n".join(missing_placeholders + [text_prompt])


def _list_replace(obj: list, old: Any, new: Any, n: int = 1) -> None:
    assert n > 0

    idx = 0
    while n != 0 and idx != len(obj):
        if obj[idx] == old:
            obj[idx] = new
            n -= 1
        idx += 1


def _full_multimodal_text_prompt_interleave(placeholder_counts: Dict[str, int],
                                            texts: List[str]) -> str:
    for placeholder, n in placeholder_counts.items():
        _list_replace(texts, IMAGE_PLACEHOLDER, placeholder, n)
    
    return "\n".join(texts)

..............

def _parse_chat_message_content_parts(
    role: str,
    parts: Iterable[ChatCompletionContentPartParam],
    mm_tracker: BaseMultiModalItemTracker,
    *,
    wrap_dicts: bool,
    interleave: bool = True
) -> List[ConversationMessage]:
    content = list[_ContentPart]()

    mm_parser = mm_tracker.create_parser()

    for part in parts:
        parse_res = _parse_chat_message_content_part(
            part,
            mm_parser,
            wrap_dicts=wrap_dicts,
            interleave=interleave
        )
        if parse_res:
            content.append(parse_res)

    if wrap_dicts:
        # Parsing wraps images and texts as interleaved dictionaries
        return [ConversationMessage(role=role,
                                    content=content)]  # type: ignore
    texts = cast(List[str], content)
    
    mm_placeholder_counts = mm_parser.mm_placeholder_counts()
    if mm_placeholder_counts:
        text_prompt = _get_full_multimodal_text_prompt(mm_placeholder_counts,
                                                       texts,
                                                       interleave=interleave)
    else:
        text_prompt = "\n".join(texts)

    return [ConversationMessage(role=role, content=text_prompt)]


def _parse_chat_message_content_part(
    part: ChatCompletionContentPartParam,
    mm_parser: BaseMultiModalContentParser,
    *,
    wrap_dicts: bool,
    interleave: bool
) -> Optional[_ContentPart]:
....
if part_type == "image_url":
        str_content = cast(str, content)
        mm_parser.parse_image(str_content)
        return {'type': 'image'} if wrap_dicts else (IMAGE_PLACEHOLDER if interleave else None)
....

```

If these modifications are ok, I'm ready to submit a properly prepared PR

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by Dekakhrone**
Current approach stacks all image tokens at the beginning of the prompt for the 'string' chat template content format, which may negatively affect the performance of some models. This PR introduces an option to enable fully interleaved support for such prompts.

FIX [#12885](https://github.com/vllm-project/vllm/issues/12885)

### CODE REVIEW

File: vllm/config.py [hunk 1]
```diff
@@ -165,6 +165,10 @@ class ModelConfig:
             the model name will be the same as `model`.
         limit_mm_per_prompt: Maximum number of data items per modality
             per prompt. Only applicable for multimodal models.
+        interleave_mm_strings: Enable fully interleaved support for multimodal 
+            (image + text) prompts, while using 
+            chat-template-content-format=string.
+            Defaults to False.
```
Comment (DarkLight1337) [inline vllm/config.py:171]: Could you explain here under what circumstances should this be set to `False`?

Comment (Dekakhrone) [reply vllm/config.py:171]: I used 'False' as the default value to preserve the original multimodal prompt processing, so the new functionality is only activated when explicitly intended by the user.

Comment (DarkLight1337) [reply vllm/config.py:171]: I think the previous behavior is actually unintentional. @ywang96 correct me if I'm wrong, but we should follow the order in which the chat template inserts the prompt segments (whether it be text or image).

Comment (Dekakhrone) [reply vllm/config.py:171]: @DarkLight1337 any updates?

Comment (DarkLight1337) [reply vllm/config.py:171]: Ok let's just default to `False` to play safe then. I also noticed that you only added this support for image inputs, would it be feasible to make it work with other multimodal inputs as well?

Comment (DarkLight1337) [reply vllm/config.py:171]: The placeholder logic is the same as that for image models.

Comment (Dekakhrone) [reply vllm/config.py:171]: Got it! I’ll work on it next week then.

Comment (DarkLight1337) [reply vllm/config.py:171]: Thanks and sorry for the delay

Comment (ywang96) [reply vllm/config.py:171]: @Dekakhrone Sorry for the late response, but the reason this was done because back then most VLMs don't have a chat template to convert `message` format to string, and since most VLMs append the placeholder to the front (Llava, Blip2, etc) we decided to do the same thing when the chat templates can only deal with string input.

**Comment by Dekakhrone** [comment]
@DarkLight1337 @ywang96 hi! Sorry for the delay - been quite busy. I've rebased my branch and added support for other modalities

File: tests/entrypoints/test_chat_utils.py [hunk 2]
```diff
@@ -555,6 +555,207 @@ def test_parse_chat_messages_multiple_images_uncommon_input(
     _assert_mm_data_is_image_input(mm_data, 2)
 
 
+def test_parse_chat_messages_multiple_images_interleave(
+    phi3v_model_config,
+    phi3v_tokenizer,
+    image_url,
+):
+    phi3v_model_config.interleave_mm_strings = True
```
Comment (DarkLight1337) [inline tests/entrypoints/test_chat_utils.py:563]: I think this may break subsequent tests since you're mutating the fixture

File: tests/entrypoints/test_chat_utils.py [hunk 3]
```diff
@@ -555,6 +555,207 @@ def test_parse_chat_messages_multiple_images_uncommon_input(
...
+        "role":
+        "user",
+        "content":
+        "I need you to compare this image\n<|image_1|>\nand this one\n<|image_2|>\n"  # noqa: E501
+        "Do they have differences?"
+    }]
+    _assert_mm_data_is_image_input(await mm_data, 2)
+
+
+def test_parse_chat_messages_multiple_images_multimple_messages_interleave(
```
Comment (DarkLight1337) [inline tests/entrypoints/test_chat_utils.py:653]:
```suggestion
def test_parse_chat_messages_multiple_images_multiple_messages_interleave(
```

File: tests/entrypoints/test_chat_utils.py [hunk 4]
```diff
@@ -555,6 +555,207 @@ def test_parse_chat_messages_multiple_images_uncommon_input(
...
+                        "Do they have differences?"
+                    },
+                ]
+            }],
+            phi3v_model_config,
+            phi3v_tokenizer,
+            content_format="string",
+        )
+
+
```
Comment (DarkLight1337) [inline tests/entrypoints/test_chat_utils.py:758]: Can you add a test case for different modalities as well?

File: vllm/engine/arg_utils.py [hunk 5]
```diff
@@ -659,6 +660,11 @@ def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
                   'e.g.: `image=16,video=2` allows a maximum of 16 '
                   'images and 2 videos per prompt. Defaults to 1 for '
                   'each modality.'))
+        parser.add_argument(
+            "--interleave-mm-strings",
+            action='store_true',
+            help=('Enable fully interleaved support for multimodal prompts, '
+                  'while using chat-template-content-format=string'))
```
Comment (DarkLight1337) [inline vllm/engine/arg_utils.py:667]:
```suggestion
                  'while using --chat-template-content-format=string'))
```

**Review by DarkLight1337 — COMMENTED** [review]
Thanks for updating, here's a new round of comments. No rush!

File: vllm/entrypoints/chat_utils.py [hunk 6]
```diff
@@ -43,6 +44,20 @@
...
+    VIDEO = "video"
+
+
+MODALITY_PLACEHOLDERS_MAP = {
+    Modality.IMAGE: "<##IMAGE##>",
+    Modality.AUDIO: "<##AUDIO##>",
+    Modality.VIDEO: "<##VIDEO##>",
+}
+
+
```
Comment (DarkLight1337) [inline vllm/entrypoints/chat_utils.py:60]: Do we really need to make this an enum?

File: vllm/entrypoints/chat_utils.py [hunk 7]
```diff
@@ -535,13 +552,15 @@ def _placeholder_str(self, modality: ModalityStr,
                                               hf_config.video_token_index)
             raise TypeError(f"Unknown {modality} model type: {model_type}")
         else:
-            raise TypeError(f"Unknown modality: {modality}")
+            raise TypeError(f"Unknown {modality}")
```
Comment (DarkLight1337) [inline vllm/entrypoints/chat_utils.py:555]: Looks unnecessary

File: vllm/entrypoints/chat_utils.py [hunk 8]
```diff
@@ -845,15 +898,22 @@ def _get_full_multimodal_text_prompt(placeholder_counts: dict[str, int],
         placeholder_counts[placeholder] -= text_prompt.count(placeholder)
 
         if placeholder_counts[placeholder] < 0:
+            logger.error(
+                "Placeholder count is negative! "
+                "Ensure that the 'interleave_strings' flag is disabled "
+                "(current value: %s) "
+                "when manually placing image placeholders.", interleave_strings
+            )
+            logger.debug(text_prompt)
```
Comment (DarkLight1337) [inline vllm/entrypoints/chat_utils.py:907]:
```suggestion
            logger.debug("Input prompt: %s", text_prompt)
```

File: tests/entrypoints/test_chat_utils.py (hunk 4 above)
Comment (Dekakhrone) [reply tests/entrypoints/test_chat_utils.py:758]: Sure! I understand that I need to use [assets](https://github.com/vllm-project/vllm/tree/main/vllm/assets) by analogy with images?

File: vllm/entrypoints/chat_utils.py (hunk 6 above)
Comment (Dekakhrone) [reply vllm/entrypoints/chat_utils.py:60]: I switched to enum because of the frequently occurring string constants - it seemed to me that it would be more convenient and safer.

Comment (DarkLight1337) [reply vllm/entrypoints/chat_utils.py:60]: We already have type checking in place to enforce this, so it should be fine to use string literals

File: vllm/entrypoints/chat_utils.py (hunk 7 above)
Comment (Dekakhrone) [reply vllm/entrypoints/chat_utils.py:555]: Github precommit check will reject it unless we cover all edge cases

File: tests/entrypoints/test_chat_utils.py (hunk 4 above)
Comment (DarkLight1337) [reply tests/entrypoints/test_chat_utils.py:758]: Yeah those two should be enough already

Comment (DarkLight1337) [reply tests/entrypoints/test_chat_utils.py:758]: Alternatively we can use qwen2.5-omni which supports all three

File: vllm/entrypoints/chat_utils.py (hunk 7 above)
Comment (DarkLight1337) [reply vllm/entrypoints/chat_utils.py:555]: I'm referring to the change in the error message

Comment (Dekakhrone) [reply vllm/entrypoints/chat_utils.py:555]: Ah, got it. Yes, I'll revert the changes along with dropping the enum

File: requirements/test.in [hunk 9]
```diff
@@ -34,7 +34,7 @@ opencv-python-headless >= 4.11.0 # required for video test
 datamodel_code_generator # required for minicpm3 test
 lm-eval[api]==0.4.8 # required for model evaluation test
 mteb>=1.38.11, <2 # required for mteb test
-transformers==4.51.3
+transformers==4.52.3
```
Comment (DarkLight1337) [inline requirements/test.in:37]: I have a separate PR to update transformers version so let's revert the changes to requirements here

File: vllm/entrypoints/chat_utils.py [hunk 10]
```diff
@@ -1042,6 +1096,7 @@ def _parse_chat_message_content_part(
     mm_parser: BaseMultiModalContentParser,
     *,
     wrap_dicts: bool,
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/entrypoints/chat_utils.py:1098]: Maybe better to rename `interleave_strings` to `wrap_dicts`

File: requirements/test.in (hunk 9 above)
Comment (Dekakhrone) [reply requirements/test.in:37]: Qwen2.5-Omni was added in 4.52 library version, so the full multimodal test won't work on the old version. Should we revert it anyway?

**Comment by DarkLight1337** [comment]
The other PR has been merged, can you update this PR accordingly?

**Comment by Dekakhrone** [comment]
@DarkLight1337 new update is ready

File: requirements/test.txt [hunk 11]
```diff
@@ -31,6 +31,10 @@ argcomplete==3.5.1
     # via datamodel-code-generator
 arrow==1.3.0
     # via isoduration
+async-timeout==5.0.1
```
Comment (DarkLight1337) [inline requirements/test.txt]: Can you update this file using Python 3.12? (Or just revert the changes to this file since you didn't add any dependencies to the code)

Comment (Dekakhrone) [reply requirements/test.txt]: Sure, just give me a couple of days

**Review by DarkLight1337 — APPROVED** [review]
LGTM now, thanks for your hard work!

**Comment by Dekakhrone** [comment]
Great! Thank you for your patience!
